src/tools/wasm-emscripten-finalize.cpp

```diff
@@ -233,7 +240,13 @@ int main(int argc, const char* argv[]) {
     }
   }
 
-  generator.generateDynCallThunks();
+  if (wasi) {
+    // Export a standard wasi "_start" method.
+    generator.exportWasiStart();
```

Comment:
I think a better approach might be to modify the linker flags than libraries when build with WASI=1 in emscripten.

So only add `--entry=main` when WASI=1 is not set (_start is the default in lld).

And also link with crt1.c which would be a file which has _start which calls main (Perhaps a simplified version of https://github.com/CraneStation/wasi-libc/blob/master/libc-bottom-half/crt/crt1.c).

Reply:
Interesting, but I'm not sure how this would work? It looks like `--entry=main` errors when there is no main. We don't know in emcc.py whether it's ok to pass that flag or not, as we don't know if there is a main or not (might be just some exported functions).

Maybe emcc is too lenient in that you can build either a program with a main() or a library or a plugin, etc., without explicitly passing a flag to say which of them it is, but I think it's useful to stay compatible with that. Maybe I'm missing a solution here though?

Reply:
In the long run i think it should be emscripten that provides `_start` .. which then calls `main`.   So binaryen would never need to do it.     In the short, term if you want to add this wrapping in binaryen I'm not going to block it.